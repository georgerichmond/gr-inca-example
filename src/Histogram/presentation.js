// @flow
import React from 'react';
import { Axis, Chart, Geom, Tooltip } from 'bizcharts';
import dl from 'datalib';
import moment from 'moment-timezone';
import type { ActivityPeriod } from '../../types';
import getInterval from './getInterval';
import { Spin } from 'antd';
import { map, prop } from 'ramda';

const tooltipColour = 'white';
const tooltipBackgroundColour = '#283c4f';

const commonChartProps = {
  forceFit: true,
  height: 100
};

const getMax = data => {
  if (Math.max(...map(prop('y'), data)) < 10) return 10;
};

const getChartProps = ({ data, name, activityPeriod }) => ({
  ...commonChartProps,
  padding: 'auto',
  scale: {
    x: {
      type: 'time',
      formatter: milliseconds => {
        const format = getInterval(activityPeriod) === '1h' ? 'HH' : 'DD';
        return moment(new Date(milliseconds).toISOString()).format(format);
      },
      tickCount: 6,
      min: activityPeriod.from,
      max: activityPeriod.to
    },
    y: {
      alias: name,
      tickCount: 3,
      formatter: dl.format.auto.number('s'),
      min: 0,
      max: getMax(data)
    }
  }
});

const getXAxisProps = ({ activityPeriod }) => ({
  label: {
    autoRotate: false
  },
  tickLine: { length: 2 }
});

type Props = {
  name: string,
  data: [{ x: string, y: number }],
  metric: string,
  activityPeriod: ActivityPeriod
};

export const HistogramLoading = () => (
  <Spin>
    <Chart {...commonChartProps} />
  </Spin>
);

export const HistogramPresentation = ({
  name,
  metric,
  data,
  activityPeriod
}: Props) => (
  <div data-test={`${metric}-histogram`}>
    <Chart data={data} {...getChartProps({ name, data, activityPeriod })}>
      <Axis name="x" {...getXAxisProps({ activityPeriod })} />
      <Axis name="y" />
      <Tooltip
        crosshairs={{ type: 'line' }}
        g2-tooltip={{
          marginTop: '-75px',
          marginLeft: '-25px',
          color: tooltipColour,
          backgroundColor: tooltipBackgroundColour,
          border: 'none',
          boxShadow: 'none'
        }}
      />
      <Geom
        size={4}
        type="interval"
        position="x*y"
        tooltip={[
          'x*y',
          (dateStr, value) => ({
            name,
            title: moment(dateStr).format('ddd Do MMM HH:mm'),
            value: value.toLocaleString()
          })
        ]}
      />
    </Chart>
  </div>
);
