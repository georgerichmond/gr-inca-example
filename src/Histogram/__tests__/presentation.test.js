import React from 'react';
import { shallow } from 'enzyme';
import { HistogramPresentation } from '../presentation';

describe('HistogramPresentation', () => {
  describe('when given data', () => {
    test('it renders correctly', () => {
      const props = {
        data: [
          { x: '2018-09-30T00:00:00Z', y: 9903 },
          { x: '2018-09-30T01:00:00Z', y: 2506 }
        ],
        name: 'Unique visitors',
        metric: 'uniqueVisitors',
        activityPeriod: {
          from: '2018-11-04T00:00:00.000Z',
          to: '2018-11-05T00:00:00.000Z'
        }
      };
      const tree = shallow(<HistogramPresentation {...props} />);
      expect(tree).toMatchSnapshot();
    });
  });
});
