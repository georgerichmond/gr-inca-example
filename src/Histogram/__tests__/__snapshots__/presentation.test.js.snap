// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HistogramPresentation when given data it renders correctly 1`] = `
<div
  data-test="uniqueVisitors-histogram"
>
  <BChart
    data={
      Array [
        Object {
          "x": "2018-09-30T00:00:00Z",
          "y": 9903,
        },
        Object {
          "x": "2018-09-30T01:00:00Z",
          "y": 2506,
        },
      ]
    }
    forceFit={true}
    height={100}
    padding="auto"
    scale={
      Object {
        "x": Object {
          "formatter": [Function],
          "max": "2018-11-05T00:00:00.000Z",
          "min": "2018-11-04T00:00:00.000Z",
          "tickCount": 6,
          "type": "time",
        },
        "y": Object {
          "alias": "Unique visitors",
          "formatter": [Function],
          "max": undefined,
          "min": 0,
          "tickCount": 3,
        },
      }
    }
  >
    <TypedComponent
      label={
        Object {
          "autoRotate": false,
        }
      }
      name="x"
      tickLine={
        Object {
          "length": 2,
        }
      }
    />
    <TypedComponent
      name="y"
    />
    <TypedComponent
      crosshairs={
        Object {
          "type": "line",
        }
      }
      g2-tooltip={
        Object {
          "backgroundColor": "#283c4f",
          "border": "none",
          "boxShadow": "none",
          "color": "white",
          "marginLeft": "-25px",
          "marginTop": "-75px",
        }
      }
    />
    <TypedComponent
      position="x*y"
      size={4}
      tooltip={
        Array [
          "x*y",
          [Function],
        ]
      }
      type="interval"
    />
  </BChart>
</div>
`;
